libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,70 +134,148 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
+                if (DateTime.UtcNow > refreshTime)
+                {
+                    // set new refresh time (this keeps only one outstanding refresh Task at a time)
+                    autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
+
+                    // background task to refresh the token
+                    // NOTE: This is not awaited, but is observed with the ContinueWith() clause.  
+                    // It is gated by the AutoRefreshTimes[] array
+                    RefreshTokenAsync()
+                        .ContinueWith(task =>
+                        {
+                            // observe the background task and put in cache when done
+                            if (task.Status == TaskStatus.RanToCompletion)
+                            {
+                                // update the cache with completed task so all new requests will get it
+                                tokenTaskCache[TokenCacheKey] = task;
+                            }
+                            else
+                            {
+                                // it failed, shorten the refresh time for another task to try again in a 30s
+                                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + TimeSpan.FromSeconds(30);
+                            }
+                        });
+                }
+            }
+        }
+
+        /// <summary>
+        /// Do not call this except from GetTokenRefreshAsync() 
+        /// </summary>
+        /// <returns></returns>
+        private Task<OAuthResponse> _getCurrentTokenTask(bool forceRefresh)
+        {
+            Task<OAuthResponse> oAuthTokenTask;
+
+            // if there is not a task or we are forcing it
+            if (forceRefresh || tokenTaskCache.TryGetValue(TokenCacheKey, out oAuthTokenTask) == false)
             {
-                _cache[TokenCacheKey] = token;
+                // create it
+                oAuthTokenTask = RefreshTokenAsync();
+                tokenTaskCache[TokenCacheKey] = oAuthTokenTask;
+
+                // set initial refresh time
+                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
+            }
+            // if task is in faulted or canceled state then replace it with another attempt
+            else if (oAuthTokenTask.IsFaulted || oAuthTokenTask.IsCanceled)
+            {
+                oAuthTokenTask = RefreshTokenAsync();
+                tokenTaskCache[TokenCacheKey] = oAuthTokenTask;
+
+                // set initial refresh time
+                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
             }
 
-            return token.access_token;
+            return oAuthTokenTask;
         }
 
         private bool ShouldSetToken(HttpRequestMessage request)
         {
-            if (IsTrustedUrl(request.RequestUri))
+            if (TrustedUri(request.RequestUri))
```

Comment:
These are URL's, not URIs.

Reply:
... I think. :)

Reply:
I brought your code forward

---
In reply to: [198677598](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198677598) [](ancestors = 198677598)